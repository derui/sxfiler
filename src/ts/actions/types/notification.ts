// actions for notification

enum ActionTypes {
  timeout = "notification_timeout",
}

export default ActionTypes;
