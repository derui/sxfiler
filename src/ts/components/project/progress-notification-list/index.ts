export * from "./progress-notification-list";
