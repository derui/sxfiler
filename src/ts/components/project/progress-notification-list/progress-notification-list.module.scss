@import "../../ui/commom-style/_colors.scss";
@import "../../ui/commom-style/_variables.scss";
@import "../../ui/commom-style/reset.module.css";

.root {
  composes: list-base from "../../ui/list/list.module.scss";
}
