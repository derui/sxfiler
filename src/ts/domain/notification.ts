
// type of notification
export enum NotificationType {
  OneShot = 'oneshot',
  Progress = 'progress',
}

export enum Level {
  Info = 'info',
  Warning = 'warning',
  Error = 'error'
}

export interface OneShotBody {
  kind: "oneshot";
  message: string;
}

export interface ProgressBody {
  kind: "progress";
  process: string;
  current: number;
  target: number;
}

export type Body = OneShotBody | ProgressBody

// notification from server
export class Notification {
  constructor(
    public readonly id: string,
    public readonly level: Level,
    public readonly body: Body
  ) {}
}
