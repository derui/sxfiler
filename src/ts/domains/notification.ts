// type of notification
export enum NotificationKind {
  Message = "message",
  Progress = "progress",
}

export enum Level {
  Info = "info",
  Warning = "warning",
  Error = "error",
}

export type MessageBody = {
  readonly message: string;
};

export type ProgressBody = {
  readonly process: string;
  readonly current: number;
  readonly target: number;
};

export type Body = MessageBody | ProgressBody;

export type MessageNotification = {
  readonly id: string;
  readonly kind: "message";
  readonly level: Level;
  readonly body: MessageBody;
};

export type ProgressNotification = {
  readonly id: string;
  readonly kind: "progress";
  readonly level: Level;
  readonly body: ProgressBody;
};

export type Notification = MessageNotification | ProgressNotification;

/**
 * factory function to create one-shot notification
 * @param id
 * @param level
 * @param message
 */
export const createMessage = (id: string, level: Level, message: string): MessageNotification => {
  return { kind: NotificationKind.Message, id, level, body: { message } };
};

/**
 * factory function to create progress notification
 * @param id
 * @param level
 * @param progress
 */
export const createProgress = (
  id: string,
  level: Level,
  progress: { process: string; current: number; target: number }
): Notification => {
  return {
    kind: NotificationKind.Progress,
    id,
    level,
    body: { ...progress },
  };
};
