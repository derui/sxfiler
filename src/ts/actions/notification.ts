import { Notification } from "../domains/notification";
import { AppAction } from "./type";

export enum ActionTypes {
  timeout = "notification_timeout",
  notify = "notification_notify",
  remove = "notification_remove",
}

type TimeoutAction = AppAction<ActionTypes.timeout, { notificationId: string }>;
type NotifyAction = AppAction<ActionTypes.notify, { notification: Notification }>;
type RemoveAction = AppAction<ActionTypes.remove, { notificationId: string }>;

export type Actions = TimeoutAction | NotifyAction | RemoveAction;

const timeout = (notificationId: string): TimeoutAction => {
  return { type: ActionTypes.timeout, notificationId };
};

const notify = (notification: Notification): NotifyAction => {
  return { type: ActionTypes.notify, notification };
};

const remove = (notificationId: string): RemoveAction => {
  return { type: ActionTypes.remove, notificationId };
};

export const actions = { timeout, notify, remove };
