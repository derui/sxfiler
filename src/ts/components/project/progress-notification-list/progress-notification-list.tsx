import * as React from "react";
import { styled } from "@/components/theme";

import { ProgressNotification } from "@/domains/progress-notification";
import * as List from "@/components/ui/list";
import { Component as Item } from "@/components/project/progress-notification-item";

export type Props = {
  notifications: ProgressNotification[];
};

const StyledList = styled(List.Component)`
  ${List.style}
`;

export const Component: React.FC<Props> = ({ notifications }) => {
  const items = notifications.map(v => <Item key={v.id} body={v.body} />);
  return <StyledList>{items}</StyledList>;
};
