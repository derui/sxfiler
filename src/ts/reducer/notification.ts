// reducers for notification
import { Actions } from "../actions";
import ActionTypes from "../actions/types/notification";
import { empty, State } from "../types/store-state/notification";

export function reducer(state: State = empty(), action: Actions): State {
  switch (action.type) {
    case ActionTypes.timeout:
      return { ...state, notifications: state.notifications.remove(action.notificationId) };
  }
  return state;
}

export default reducer;
