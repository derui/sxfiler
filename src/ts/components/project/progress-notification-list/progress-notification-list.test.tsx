import * as React from "react";
import renderer from "react-test-renderer";

import * as N from "../../../domains/notification";
import { Component as T } from "./progress-notification-list";

describe("Project", () => {
  describe("Progress Notification List", () => {
    it("can render correctly", () => {
      const notifications = [
        N.createProgress("id", N.Level.Info, {
          process: "process",
          current: 10,
          target: 100,
        }),
      ];
      const tree = renderer.create(<T notifications={notifications} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("can render with empty list of notification", () => {
      const tree = renderer.create(<T notifications={[]} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("should render progresses", () => {
      const notifications = [
        N.createProgress("id", N.Level.Info, {
          process: "foo",
          current: 9,
          target: 10,
        }),
        N.createProgress("id2", N.Level.Warning, {
          process: "bar",
          current: 10,
          target: 11,
        }),
      ];
      const tree = renderer.create(<T notifications={notifications} />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
