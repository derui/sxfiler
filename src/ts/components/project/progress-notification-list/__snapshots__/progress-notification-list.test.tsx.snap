// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Progress Notification List can render correctly 1`] = `
<div
  className="sc-ifAKCX ipbtqc"
  role="list"
>
  <div
    aria-selected={false}
    className="sc-bdVaJa gneklD"
    role="listitem"
  >
    <span
      className="sc-bwzfXH cJKlyb"
    >
      process
    </span>
    <div
      className="sc-htpNat hkuyqD"
    >
      <div
        className="sc-bxivhb MqTMw"
      />
    </div>
  </div>
</div>
`;

exports[`Project Progress Notification List can render with empty list of notification 1`] = `
<div
  className="sc-ifAKCX ipbtqc"
  role="list"
/>
`;

exports[`Project Progress Notification List should render progresses 1`] = `
<div
  className="sc-ifAKCX ipbtqc"
  role="list"
>
  <div
    aria-selected={false}
    className="sc-bdVaJa gneklD"
    role="listitem"
  >
    <span
      className="sc-bwzfXH cJKlyb"
    >
      foo
    </span>
    <div
      className="sc-htpNat hkuyqD"
    >
      <div
        className="sc-bxivhb jQexdE"
      />
    </div>
  </div>
  <div
    aria-selected={false}
    className="sc-bdVaJa gneklD"
    role="listitem"
  >
    <span
      className="sc-bwzfXH cJKlyb"
    >
      bar
    </span>
    <div
      className="sc-htpNat hkuyqD"
    >
      <div
        className="sc-bxivhb iKWdPW"
      />
    </div>
  </div>
</div>
`;
